scripts/serve_header/README.md

```diff
@@ -0,0 +1,85 @@
+serve_header.py
+===============
+
+Serves the `single_include/nlohmann/json.hpp` header file over HTTP(S).
+
+The header file is automatically amalgamated on demand.
+
+![serve_header.py demo](https://user-images.githubusercontent.com/320854/165481390-d65a5180-00d3-443b-b2f1-ea227316309f.png)
+
+## Prerequisites
+
+> **Note:** All file locations are relative to the project root folder and all commands are assumed to be executed from the same root folder.
+
+1. Make sure these Python packages are installed.
+    ```
+    PyYAML
+    watchdog
+    ```
+    (see `scripts/serve_header/requirements.txt`)
+
+2. To serve the header over HTTPS (which is required by Compiler Explorer at this time), a certificate is needed.
+   The recommended method for creating a locally-trusted certificate is to use [`mkcert`](https://github.com/FiloSottile/mkcert).
+   - Install the `mkcert` certificate authority into your trust store(s):
+     ```
+     $ mkcert -install
+     ```
+   - Create a certificate for `localhost`:
+     ```
+     $ mkcert localhost
+     ```
+
+## Usage
+
+`serve_header.py` has a builtin default configuration that will serve the header from the project folder it is homed in.
+It expects to find the certificate `localhost.pem` and the private key `localhost-key.pem` in the projects root directory.
+
+To start serving the `json.hpp` header file at `https://localhost:8443/json.hpp`, run this command:
+```
+$ make serve_header
+```
+
+Open [Compiler Explorer](https://godbolt.org/) and try it out:
+```cpp
+#include <https://localhost:8443/json.hpp>
+using namespace nlohmann;
+
+#include <iostream>
+
+int main() {
+    // these macros are dynamically injected into the header file
+    std::cout << JSON_BUILD_TIME << " (" << JSON_BUILD_COUNT << ")\n";
+
+    return 0;
+}
+```
+
+> `serve_header.py` dynamically injects the macros `JSON_BUILD_COUNT` and `JSON_BUILD_TIME` into the served header. By comparing build count or time output from the compiled program with the output from `serve_header.py`, one can be reasonably sure the compiled code uses the expected revision of the header file.
+
+## Configuration
+
+`serve_header.py` will try to read a configuration file `serve_header.yml` and will fallback on builtin defaults, if the file cannot be read.
```

Comment:
The comma is not necessary, but could go after the second filename.

What directory should the configuration file be in?  It's unclear to me.  Should it be in the directory with the script, or the top level directory that was checked out from github, or the parent directory of that checkout?

`fallback` is a noun, the verb form is `fall back`, and `built-in` is hyphenated.

> `serve_header.py` will try to read a configuration file `serve_header.yml`, and will fall back on built-in defaults if the file cannot be read.

Reply:
> The comma is not necessary, but could go after the second filename.

Noted.

> What directory should the configuration file be in? It's unclear to me. Should it be in the directory with the script, or the top level directory that was checked out from github, or the parent directory of that checkout?

The top level directory aka. project root. There's a note about all files and commands being relative to that directory. I wanted to avoid duplicating it everywhere ... seems to have been a mistake.

> `fallback` is a noun, the verb form is `fall back`,

Noted.

> and `built-in` is hyphenated.

When I looked that up a few months ago, builtin was deemed an increasingly more acceptable way of spelling it. I've adopted that spelling since then.
But that's not a hill I want to or need to die on. Your call. ;-)
 
> > `serve_header.py` will try to read a configuration file `serve_header.yml`, and will fall back on built-in defaults if the file cannot be read.

Reply:
> When I looked that up a few months ago, builtin was deemed an increasingly more acceptable way of spelling it. I've adopted that spelling since then.

Fine by me then.

Reply:
>>What directory should the configuration file be in? It's unclear to me. Should it be in the directory with the script, or the top level directory that was checked out from github, or the parent directory of that checkout?

>The top level directory aka. project root. There's a note about all files and commands being relative to that directory. I wanted to avoid duplicating it everywhere ... seems to have been a mistake.

So in the case of multiple checkouts (manual or via git worktree) under a single parent directory, you can put it in the root of just one of the checkouts, and then have the `root: ..` to let it find the ones in the sibling directories?